common/src/main/java/org/apache/uniffle/common/netty/protocol/SendShuffleDataRequest.java

```diff
@@ -0,0 +1,153 @@
+    this.requireId = requireId;
+    this.partitionToBlocks = partitionToBlocks;
+    this.timestamp = timestamp;
+  }
+
+  @Override
+  public Type type() {
+    return Type.SEND_SHUFFLE_DATA_REQUEST;
+  }
+
+  @Override
+  public int encodedLength() {
+    int encodeLength = Long.BYTES + ByteBufUtils.encodedLength(appId) + Integer.BYTES + Long.BYTES + Integer.BYTES;
+    for (Map.Entry<Integer, List<ShuffleBlockInfo>> entry : partitionToBlocks.entrySet()) {
+      encodeLength += 2 * Integer.BYTES;
+      for (ShuffleBlockInfo sbi : entry.getValue()) {
+        encodeLength += Encoders.encodeLengthOfShuffleBlockInfo(sbi);
+      }
+    }
+    return encodeLength + Long.BYTES;
+  }
+
+  @Override
+  public void encode(ByteBuf buf) {
+    buf.writeLong(requestId);
+    ByteBufUtils.writeLengthAndString(buf, appId);
+    buf.writeInt(shuffleId);
+    buf.writeLong(requireId);
+    encodePartitionData(buf);
+    buf.writeLong(timestamp);
+  }
+
+  private static Map<Integer, List<ShuffleBlockInfo>> decodePartitionData(ByteBuf byteBuf) {
+    Map<Integer, List<ShuffleBlockInfo>> partitionToBlocks = Maps.newHashMap();
+    int lengthOfPartitionData = byteBuf.readInt();
+    for (int i = 0; i < lengthOfPartitionData; i++) {
+      int partitionId = byteBuf.readInt();
+      int lengthOfShuffleBlocks = byteBuf.readInt();
+      List<ShuffleBlockInfo> shuffleBlockInfoList = Lists.newArrayList();
+      for (int j = 0; j < lengthOfShuffleBlocks; j++) {
+        shuffleBlockInfoList.add(Decoders.decodeShuffleBlockInfo(byteBuf));
+      }
+      partitionToBlocks.put(partitionId, shuffleBlockInfoList);
+    }
+    return partitionToBlocks;
+  }
+
+  public static SendShuffleDataRequest decode(ByteBuf byteBuf) {
+    long requestId = byteBuf.readLong();
+    String appId = ByteBufUtils.readLengthAndString(byteBuf);
+    int shuffleId = byteBuf.readInt();
+    long requireId = byteBuf.readLong();
+    Map<Integer, List<ShuffleBlockInfo>> partitionToBlocks = decodePartitionData(byteBuf);
+    long timestamp = byteBuf.readLong();
+    return new SendShuffleDataRequest(requestId, appId, shuffleId, requireId, partitionToBlocks, timestamp);
+  }
+
+  private void encodePartitionData(ByteBuf buf) {
+    buf.writeInt(partitionToBlocks.size());
+    for (Map.Entry<Integer, List<ShuffleBlockInfo>> entry : partitionToBlocks.entrySet()) {
+      buf.writeInt(entry.getKey());
+      buf.writeInt(entry.getValue().size());
+      for (ShuffleBlockInfo sbi : entry.getValue()) {
+        Encoders.encodeShuffleBlockInfo(sbi, buf);
+      }
+    }
+  }
+
+  public long getRequestId() {
+    return requestId;
+  }
+
+  public String getAppId() {
+    return appId;
+  }
+
+  public int getShuffleId() {
+    return shuffleId;
+  }
+
+  public long getRequireId() {
+    return requireId;
+  }
+
+  public Map<Integer, List<ShuffleBlockInfo>> getPartitionToBlocks() {
+    return partitionToBlocks;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
Why do we need to override the method ``equal?

Reply:
for test, see SendShuffleDataRequestTest.java.

Reply:
> for test, see SendShuffleDataRequestTest.java.

We can create another method like `compareSendShuffleDataRequest` in the test code.

Reply:
I worried about performance of this method `equal`.

Reply:
Ok.